src/main/java/mate.academy/lib/Injector.java

```diff
@@ -1,13 +1,81 @@
     private static final Injector injector = new Injector();
+    private final Map<Class<?>, Object> instances = new HashMap<>();
+    private final Map<Class<?>, Class<?>> interfaceImplementations = Map.of(
+            ProductService.class, ProductServiceImpl.class,
+            FileReaderService.class, FileReaderServiceImpl.class,
+            ProductParser.class, ProductParserImpl.class
+    );
 
     public static Injector getInjector() {
         return injector;
     }
 
-    public Object getInstance(Class<?> interfaceClazz) {
-        return null;
+    public Object getInstance(Class<?> interfaceClass) {
+        try {
+            if (instances.containsKey(interfaceClass)) {
+                return instances.get(interfaceClass);
+            }
+
+            Class<?> implementationClass = interfaceImplementations.get(interfaceClass);
+            if (implementationClass == null
+                    || !implementationClass.isAnnotationPresent(Component.class)) {
+                throw new IllegalArgumentException(
+                        "Injection failed, missing @Component annotation on the class "
+                                + (implementationClass != null
+                                ? implementationClass.getName() : "null")
+                );
+            }
+
+            Object instance = createInstance(implementationClass);
+            instances.put(interfaceClass, instance);
+            injectDependencies(instance);
+            return instance;
+        } catch (ReflectiveOperationException e) {
+            throw new RuntimeException(
+                    "Injection failed for " + interfaceClass.getName(), e
+            );
+        }
+    }
+
+    private Object createInstance(Class<?> clazz) throws ReflectiveOperationException {
+        Constructor<?>[] constructors = clazz.getDeclaredConstructors();
+        for (Constructor<?> constructor : constructors) {
+            if (constructor.isAnnotationPresent(Inject.class)) {
+                Class<?>[] parameterTypes = constructor.getParameterTypes();
+                Object[] parameters = new Object[parameterTypes.length];
+                for (int i = 0; i < parameterTypes.length; i++) {
+                    parameters[i] = getInstance(parameterTypes[i]);
+                }
+                return constructor.newInstance(parameters);
+            }
+        }
+        return clazz.getDeclaredConstructor().newInstance();
+    }
+
+    private void injectDependencies(Object instance) throws ReflectiveOperationException {
+        Field[] fields = instance.getClass().getDeclaredFields();
+        for (Field field : fields) {
+            if (field.isAnnotationPresent(Inject.class)) {
+                boolean accessible = field.canAccess(instance);
+                field.setAccessible(true);
+                Object fieldInstance = getInstance(field.getType());
+                field.set(instance, fieldInstance);
+                field.setAccessible(accessible);
```

Comment:
You are temporarily changing the accessibility of the field to inject the dependency, which is good. However, make sure to always restore the field's original accessibility state after injection to avoid potential security issues.